### PR TITLE
Refactor cloud messages

### PR DESCRIPTION
**Pull request by ramonhpr**

### CODE REVIEW

File: src/cloud.c [hunk 1]
```diff
@@ -162,8 +162,15 @@ int cloud_set_cbs(cloud_downstream_cb_t on_update,
 		  cloud_device_removed_cb_t on_removed,
 		  void *user_data)
 {
+	static const char * const fog_events[] = {
```
Comment (alla3) [inline src/cloud.c]: I would let it just a `const char *`.

File: src/cloud.c [hunk 2]
```diff
@@ -56,60 +56,83 @@
...
-		cloud_cbs.update_cb(id, list, user_data);
+		cb_handler.update_cb(id, list, user_data);
 		l_queue_destroy(list, l_free);
-	}
-
-	if (cloud_cbs.request_cb != NULL &&
-	    strcmp(AMQP_EVENT_DATA_REQUEST, routing_key) == 0) {
+		break;
+	case REQUEST_EVT:
 		return true;
```
Comment (alla3) [inline src/cloud.c]: I would change the `return true;` for a `break;`.

File: src/cloud.c [hunk 3]
```diff
@@ -56,60 +56,83 @@
...
 #define AMQP_CMD_DEVICE_UNREGISTER "device.unregister"
 
-struct cloud_callbacks {
+/* Hashmap to user callback from southbound traffic */
+static struct l_hashmap *map_event_to_cb;
+
+/* Union with all cloud callbacks */
+union cloud_cb_t {
 	cloud_downstream_cb_t update_cb;
-	cloud_downstream_cb_t request_cb;
```
Comment (alla3) [inline src/cloud.c]: Why did you remove this callback from the union?

File: src/cloud.c (hunk 2 above)
Comment (vitbaq) [reply src/cloud.c]: Use break because of the json_object_put(jso); on line 134

File: src/cloud.c (hunk 1 above)
Comment (ramonhpr) [reply src/cloud.c]: Done

File: src/cloud.c [hunk 4]
```diff
@@ -162,8 +162,15 @@ int cloud_set_cbs(cloud_downstream_cb_t on_update,
...
+	static const char * const fog_events[] = {
+		AMQP_EVENT_DATA_UPDATE,
+		AMQP_EVENT_DATA_REQUEST,
+		AMQP_EVENT_DEVICE_UNREGISTERED,
+		NULL
+	};
 	amqp_bytes_t queue_fog;
-	int err;
+	int err, i;
+
```
[the remark this answers is not in the document]
Comment (ramonhpr) [reply src/cloud.c]: Removed

File: src/cloud.c (hunk 3 above)
Comment (ramonhpr) [reply src/cloud.c]: Done

File: src/cloud.c (hunk 2 above)
Comment (ramonhpr) [reply src/cloud.c]: Done

File: src/cloud.c [hunk 5]
```diff
@@ -92,38 +104,54 @@ static bool on_cloud_receive_message(const char *exchange,
 	struct l_queue *list;
 	const char *id;
 	bool consumed = true;
+	const struct event_handler *handler;
```
Comment (alla3) [inline src/cloud.c]: As you are using a **cb_handler** maybe it would be better if this variable was declared as something like **evt_handler**

Comment (ramonhpr) [reply src/cloud.c]: Done
